File: leetcode/819.cpp
class Solution {
public:
    string getLower(string &x) {
        string s;
        for(char c : x) s += tolower(c);
        return s;
    }
    string mostCommonWord(string paragraph, vector<string>& banned) {
        map<string, int> m;
        int index = 0, maxCount=0;
        while(index<paragraph.size()) {
            while(index<paragraph.size() && !isalpha(paragraph[index]))++index;
            if(index==paragraph.size()) break;
            int start = index;
            while(index<paragraph.size() && isalpha(paragraph[index]))++index;
            int end = index;
            string s = paragraph.substr(start, end-start);
            s = getLower(s);
            if(find(banned.begin(), banned.end(), s)!=banned.end()) continue;
            m[s]++;
            maxCount=max(maxCount, m[s]);
        }
        string res;
        for(auto & i : m) {
            if(i.second == maxCount) {
                res=i.first;
                break;
            }
        }
        return res;
    }
};

File: leetcode/819.go
func toLowerRune(r rune) rune {
	if r >= 'a' && r <= 'z' {
		return r
	}
	if r >= 'A' && r <= 'z' {
		return r + 'a' - 'A'
	}
	return r
}

func toLowerString(x string) string {
	var s []rune
	for _, r := range x {
		s = append(s, toLowerRune(r))
	}
	return string(s)
}

func isDigit(b byte) bool {
	return rune(b) >= 'a' && rune(b) <= 'z' || rune(b) >= 'A' && rune(b) <= 'Z'
}

func mostCommonWord(paragraph string, banned []string) string {
	m := make(map[string]int)
	index := 0
	maxCount := 0
	for index < len(paragraph) {
		for index < len(paragraph) && !isDigit(paragraph[index]) {
			index++
		}
		start := index
		for index < len(paragraph) && isDigit(paragraph[index]) {
			index++
		}
		end := index
		flag := false
		s := toLowerString(paragraph[start:end])
		for _, x := range banned {
			if s == x {
				flag = true
				break
			}
		}
		if !flag {
			m[s]++
			if m[s] > maxCount {
				maxCount = m[s]
			}
		}
	}
	for k, v := range m {
		if v == maxCount {
			return k
		}
	}
	return ""
}